package/metrics_daemon/metrics_daemon/src/metrics_daemon.c

```diff
@@ -0,0 +1,378 @@
+  if (pdest == NULL) {
+    pdest = str; // The whole name is a file in current path
+  } else {
+    pdest++; // Skip the backslash itself.
+  }
+
+  // extract filename from file path
+  if (pdest[0] == '.') pdest++;
+  inpfile = calloc(1, MAX_FOLDER_NAME_LENGTH + 1); // Make space for the zero.
+  strncpy(inpfile, pdest, MAX_FOLDER_NAME_LENGTH); // Copy including zero.
+  return inpfile;
+}
+
+
+/**
+ * @brief function that use the full file path loop through the json object tree
+ * to update values
+ *
+ * @param process_path file path in ASCII
+ * @param root         root name, this is used to pair with the root of the json
+ * object tree
+ * @param root_len     number of chars for the root name
+ * @param json_root    root of the json object tree
+ * @return json object node that needs to be updated
+ */
+static struct json_object *loop_through_folder_name(const char *process_path,
+                                                    const char *root,
+                                                    unsigned int root_len)
+{
+  char ch = '/';
+  const char *pdest = NULL;
+  bool found_root = false;
+  pdest = strchr(process_path, ch); // get the first slash position
+  unsigned int str_len = (unsigned int)strlen(process_path);
+  if (jobj_root == NULL || pdest < process_path) {
+    return jobj_root;
+  }
+  struct json_object *json_current = jobj_root;
+  while (pdest != NULL) {
+    pdest++;
+    if (pdest > process_path + str_len) return json_current;
+    if (pdest[0] == '.') // skip the '.' on the file name
+    {
+      pdest++;
+    }
+    if (pdest > process_path + str_len) return json_current;
+    char *pdest2 = strchr(pdest, ch); // search the second slash
+    if (pdest2 != NULL) {
+      unsigned int strlen;
+      if (pdest2 <= pdest) // defensive code
+        return json_current;
+      else
+        strlen = (unsigned)(pdest2 - pdest); // cast to unsigned since the negative and zero
+                                             // case is handled by the if statement
+      if (!found_root && strncmp(pdest, root, root_len) == 0) // search the root first
+      {
+        found_root = true;
+        pdest--;
+        continue;
+      }
+      if (found_root) {
+        bool found_target = false;
+        json_object_object_foreach(json_current, key, val)
+        { // for each subfolder, search the name as key in json tree
+          if (strncmp((pdest), key, strlen - 1) == 0) // if find the folder name, continue on the
+                                                      // loop with the new current node
+          {
+            json_current = val;
+            found_target = true;
+            break;
+          }
+        }
+        if (!found_target) // if couldn't find target on the tree, return
+                           // current node to add new node
+          return json_current;
+      }
+    }
+    pdest = pdest2; // if there is still sub-folders, update the pdest to pdest2
```

Comment:
In general for naming convention, let's avoid adding 'p' as a way to indicate a pointer. Names like `pdest` and `pdest2` start to lose their meaning. If necessary, refactor into smaller helper functions so the names remain relevant.